pydantic/types.py

```diff
@@ -2429,3 +2438,216 @@ def __getattr__(self, item: str) -> Any:
+
+
+    class Pie(BaseModel):
+        time_to_cook: int
+        num_ingredients: int
+
+
+    class ApplePie(Pie):
+        fruit: Literal['apple'] = 'apple'
+
+
+    class PumpkinPie(Pie):
+        filling: Literal['pumpkin'] = 'pumpkin'
+
+
+    def get_discriminator_value(v: Any) -> str:
+        if isinstance(v, dict):
+            return v.get('fruit', v.get('filling'))
+        return getattr(v, 'fruit', getattr(v, 'filling', None))
+
+
+    class ThanksgivingDinner(BaseModel):
+        dessert: Annotated[
+            Union[
+                Annotated[ApplePie, Tag("apple")],
+                Annotated[PumpkinPie, Tag("pumpkin")],
+            ],
+            CallableDiscriminator(get_discriminator_value),
+        ]
+
+
+    apple_variation = ThanksgivingDinner.model_validate(
+        {'dessert': {'fruit': 'apple', 'time_to_cook': 60, 'num_ingredients': 8}}
+    )
+    print(repr(apple_variation))
+    # > ThanksgivingDinner(dessert=ApplePie(fruit='apple', time_to_cook=60, num_ingredients=8))
+
+    pumpkin_variation = ThanksgivingDinner.model_validate(
+        {'dessert': {'filling': 'pumpkin', 'time_to_cook': 40, 'num_ingredients': 6}}
+    )
+    print(repr(pumpkin_variation))
+    # > ThanksgivingDinner(dessert=PumpkinPie(filling='pumpkin', time_to_cook=40, num_ingredients=6))
+    ```
+
+    See the [Discriminated Unions](../api/standard_library_types.md#discriminated-unions-aka-tagged-unions)
+    docs for more details on how to use `CallableDiscriminator`s.
+    """
+
+    discriminator: Callable[[Any], Hashable]
+    custom_error_type: str | None = None
+    custom_error_message: str | None = None
+    custom_error_context: dict[str, int | str | float] | None = None
+
+    def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+        origin = _typing_extra.get_origin(source_type)
+        if not origin or not _typing_extra.origin_is_union(origin):
+            raise TypeError(f'{type(self).__name__} must be used with a Union type, not {source_type}')
+
+        original_schema = handler.generate_schema(source_type)
+        return self._convert_schema(original_schema)
+
+    def _convert_schema(self, original_schema: core_schema.CoreSchema) -> core_schema.TaggedUnionSchema:
+        if original_schema['type'] != 'union':
+            # This likely indicates that the schema was a single-item union that was simplified.
+            # In this case, we do the same thing we do in
+            # `pydantic._internal._discriminated_union._ApplyInferredDiscriminator._apply_to_root`, namely,
+            # package the generated schema back into a single-item union.
+            original_schema = core_schema.union_schema([original_schema])
+
+        tagged_union_choices = {}
+        for i, choice in enumerate(original_schema['choices']):
+            tag = None
+            if isinstance(choice, tuple):
+                choice, tag = choice
```

Comment:
Let's just make sure that this doesn't cause us any issues with "default" tags. I can't remember if we exposed this on pydantic side or just pydantic-core currently..

Reply:
investigated, and I think we can use Tag to set this as a way to improve other usability of unions in pydantic, but I think we can ignore that in this PR.